File: Observer_Pattern/go/Observer_Pattern.go
package Observer_Pattern

import "fmt"

type Object interface {
	SetStat(int)
	NotifyAll()
	Attach(Observer)
}

type Observer interface {
	Notify(string)
}

type Object1 struct {
	List []Observer
	stat int
}

type Observer1 struct{}

func (this *Object1) SetStat(s int) {
	this.stat = s
	this.NotifyAll()
}

func (this *Object1) NotifyAll() {
	s := fmt.Sprintf("from object 1 stat %d notified", this.stat)
	for _, n := range this.List {
		n.Notify(s)
	}
}

func (this *Object1) Attach(ob Observer) {
	this.List = append(this.List, ob)
}

func (this *Observer1) Notify(s string) {
	fmt.Printf("ob1 %s\n", s)
}

type Observer2 struct{}

func (this *Observer2) Notify(s string) {
	fmt.Printf("ob2 %s\n", s)
}

File: Observer_Pattern/cpp/Observer_Pattern.cc
#include <iostream>
#include <vector>
#include <stdio.h>

using namespace std;
class Observer {
public:
    virtual void Notify(string) = 0;
};

class Object {
public:
    virtual void Attach(Observer*) = 0;
    virtual void NotifyAll() = 0; 
    virtual void SetStat(int) = 0;
protected:
    vector<Observer*> list;
    int stat;
};
class Object1 : public Object {
public:
    void Attach(Observer* ob) {
        list.push_back(ob);
    }
    void NotifyAll() {
        auto i = list.begin();
        char buf[1024];
        sprintf(buf, "from object 1 stat %d notified", stat);
        while (i != list.end()) {
            (*i)->Notify(buf);
            i++;
        }
    }
    void SetStat(int s) {
        stat = s;
        NotifyAll();
    }
};

class Observer1 : public Observer {
public:
    void Notify(string s) {
        cout << "observer1 " << s <<endl;
    }
};

class Observer2 : public Observer {
public:
    void Notify(string s) {
        cout << "observer2 " << s <<endl;
    }
};

void test_observer() {
    Object* obj = new Object1;
    Observer1 ob1;
    Observer2 ob2;
    Observer2 ob3;
    obj->Attach(&ob1);
    obj->Attach(&ob2);
    obj->Attach(&ob3);

    obj->SetStat(1);
}

int main() {
    test_observer();
    return 0;
}